### PR TITLE
Added cleaned configuration properties for tokenizer with serialization - improve tokenization of XLM

### PR DESCRIPTION
**Pull request by shijie-wu**
This PR improve the tokenization of XLM. It's mostly the same as the [preprocessing](https://github.com/facebookresearch/XLM/blob/master/tools/tokenize.sh) in the original XLM. This PR also add `use_lang_emb` to config of XLM model, which makes adding the newly release [XLM-17 & XLM-100](https://github.com/facebookresearch/XLM#pretrained-cross-lingual-language-models) easier since both of them don't have language embedding.

Details on tokenization:
- Introduce API change: Changing `XLMTokenizer.tokenize(self, text)` to  `XLMTokenizer.tokenize(text, lang='en')`
- New dependency:
  - [sacremoses](https://github.com/alvations/sacremoses): port of Moses
- New optional dependencies:
  - [pythainlp](https://github.com/PyThaiNLP/pythainlp): Thai tokenizer
  - [kytea](https://github.com/chezou/Mykytea-python): Japanese tokenizer, wrapper of [KyTea](https://github.com/neubig/kytea) (Need external C++ compilation), used by the newly release XLM-17 & XLM-100
  - [jieba](https://github.com/fxsjy/jieba): Chinese tokenizer *

\* XLM used Stanford Segmenter. However, the wrapper (`nltk.tokenize.stanford_segmenter`) are slow due to JVM overhead, and it will be deprecated. Jieba is a lot faster and pip-installable. But there is some mismatch with the Stanford Segmenter. A workaround could be having an argument to allow users to segment the sentence by themselves and bypass the segmenter. As a reference, I also include `nltk.tokenize.stanford_segmenter` in this PR.

Example of tokenization difference could be found [here](https://colab.research.google.com/drive/1nY930H2dhz3IlFvDgU9ycgfm2-DpvRcT).

### CODE REVIEW

File: pytorch_transformers/tokenization_xlm.py [hunk 1]
```diff
@@ -122,16 +201,16 @@ def __init__(self, vocab_file, merges_file, unk_token="<unk>", bos_token="<s>",
...
-            self.nlp = BasicTokenizer(do_lower_case=True)
-            self.fix_text = None
+
+        # cache of sm.MosesPunctNormalizer instance
+        self.cache_moses_punct_normalizer = dict()
+        # cache of sm.MosesTokenizer instance
+        self.cache_moses_tokenizer = dict()
+        self.lang_with_custom_tokenizer = set(['zh', 'th', 'ja'])
+        # True for current supported model (v1.2.0), False for XLM-17 & 100
+        self.do_lowercase_and_remove_accent = True
```
Comment (thomwolf) [inline pytorch_transformers/tokenization_xlm.py]: So here the workflow for the user is to manually deactivate this for XLM-17 & 100 (e.g. by doing `tokenizer. do_lowercase_and_remove_accent = False` after loading the tokenizer)?

This looks a bit error-prone to me. Maybe we should have a way to setup a configuration for the provided tokenizer. I was thinking about that for Bert already (people forget to set `do_lower_case` regularly).

I'll add something high level on the base class `PreTrainedTokenizer`.

File: pytorch_transformers/tokenization_xlm.py [hunk 2]
```diff
@@ -140,6 +219,55 @@ def __init__(self, vocab_file, merges_file, unk_token="<unk>", bos_token="<s>",
...
+    def moses_pipeline(self, text, lang):
+        text = replace_unicode_punct(text)
+        text = self.moses_punct_norm(text, lang)
+        text = remove_non_printing_char(text)
+        return text
+
+    def ja_tokenize(self, text):
+        if self.ja_word_tokenizer is None:
+            try:
+                self.ja_word_tokenizer = Mykytea.Mykytea('-model %s/local/share/kytea/model.bin' % os.path.expanduser('~'))
```
Comment (thomwolf) [inline pytorch_transformers/tokenization_xlm.py]: Right now XLM is not used enough to add `KyTea` to the required dependencies.
But this workflow with a test and clear error is very nice, we can just remove `KyTea` from the dependencies.

File: pytorch_transformers/tokenization_xlm.py [hunk 3]
```diff
@@ -140,6 +219,55 @@ def __init__(self, vocab_file, merges_file, unk_token="<unk>", bos_token="<s>",
...
+                logger.error("2. autoreconf -i")
+                logger.error("3. ./configure --prefix=$HOME/local")
+                logger.error("4. make && make install")
+                import sys; sys.exit()
+        return list(self.ja_word_tokenizer.getWS(text))
+
+    def zh_tokenize(self, text):
+        if self.zh_word_tokenizer is None:
+            try:
+                self.zh_word_tokenizer = StanfordSegmenter()
```
Comment (thomwolf) [inline pytorch_transformers/tokenization_xlm.py]: Here again, XLM is not used enough to add NLTK to the required dependencies.
The workflow with a test and clear error is nice, we can just remove NLTK from the dependencies.

File: pytorch_transformers/tokenization_xlm.py [hunk 4]
```diff
@@ -187,19 +315,34 @@ def bpe(self, token):
...
+            text = lowercase_and_remove_accent(text)
+        if lang not in self.lang_with_custom_tokenizer:
+            text = self.moses_pipeline(text, lang=lang)
+            # TODO: make sure we are using `xlm-mlm-enro-1024`, since XLM-100 doesn't have this step
+            if lang == 'ro':
+                text = romanian_preprocessing(text)
+            text = self.moses_tokenize(text, lang=lang)
+        elif lang == 'th':
+            text = self.moses_pipeline(text, lang=lang)
+            text = th_word_tokenize(text)
```
Comment (thomwolf) [inline pytorch_transformers/tokenization_xlm.py]: We also won't add pythainlp to the required dependencies so there should be here a test and a detailed error message to indicate you need to install pythainlp.

File: pytorch_transformers/tokenization_xlm.py [hunk 5]
```diff
@@ -187,19 +315,34 @@ def bpe(self, token):
...
+        elif lang == 'zh':
+            # text = self.zh_tokenize(text)
+            text = ' '.join(jieba.cut(text))
```
Comment (thomwolf) [inline pytorch_transformers/tokenization_xlm.py]: Here also, we also won't add jieba to the required dependencies so there should be here a test and a detailed error message to indicate you need to install jieba.

File: pytorch_transformers/tokenization_xlm.py [hunk 6]
```diff
@@ -187,19 +315,34 @@ def bpe(self, token):
...
-        if self.fix_text is None:
-            # Using BERT's BasicTokenizer
-            text = self.nlp.tokenize(text)
-            for token in text:
-                split_tokens.extend([t for t in self.bpe(token).split(' ')])
+        if self.do_lowercase_and_remove_accent:
+            text = lowercase_and_remove_accent(text)
+        if lang not in self.lang_with_custom_tokenizer:
+            text = self.moses_pipeline(text, lang=lang)
+            # TODO: make sure we are using `xlm-mlm-enro-1024`, since XLM-100 doesn't have this step
```
Comment (thomwolf) [inline pytorch_transformers/tokenization_xlm.py]: As mentioned above, I'll add higher-level configuration to feed the tokenizer `__init__` so you can know this.

File: pytorch_transformers/tokenization_xlm.py [hunk 7]
```diff
@@ -140,6 +219,55 @@ def __init__(self, vocab_file, merges_file, unk_token="<unk>", bos_token="<s>",
...
+            punct_normalizer = sm.MosesPunctNormalizer(lang=lang)
+            self.cache_moses_punct_normalizer[lang] = punct_normalizer
+        else:
+            punct_normalizer = self.cache_moses_punct_normalizer[lang]
+        return punct_normalizer.normalize(text)
+
+    def moses_tokenize(self, text, lang):
+        if lang not in self.cache_moses_tokenizer:
+            moses_tokenizer = sm.MosesTokenizer(lang=lang)
+            self.cache_moses_tokenizer[lang] = moses_tokenizer
```
Comment (thomwolf) [inline pytorch_transformers/tokenization_xlm.py]: Nice!

File: requirements.txt [hunk 8]
```diff
@@ -9,4 +9,10 @@ requests
 # For OpenAI GPT
 regex
 # For XLNet
-sentencepiece
\ No newline at end of file
+sentencepiece
+# For XLM
+sacremoses
+pythainlp
```
Comment (thomwolf) [inline requirements.txt]: So here we'll only keep `sacremoses` as added dependencies and have detailed error messages for the others

File: setup.py [hunk 9]
```diff
@@ -55,7 +55,12 @@
                       'requests',
                       'tqdm',
                       'regex',
-                      'sentencepiece'],
+                      'sentencepiece',
+                      'sacremoses',
```
Comment (thomwolf) [inline setup.py]: Same as above, we only keep `sacremoses` as added dependency.

File: pytorch_transformers/tokenization_xlm.py [hunk 10]
```diff
@@ -187,19 +315,34 @@ def bpe(self, token):
         self.cache[token] = word
         return word
 
-    def _tokenize(self, text):
+    def _tokenize(self, text, lang='en'):
```
Comment (thomwolf) [inline pytorch_transformers/tokenization_xlm.py]: We should fix the docstring of `XLMTokenizer` which is now totally wrong:
- remove the reference to SpaCy
- explain that people need to input a `lang` argument to the `tokenize` and `encode` function.

The `encode` method currently doesn't accept `kwargs`. I will add that and make them forwarded to `tokenize` so the user can do `tokenizer.encode(text, lang='jp')` as well.

**Review by thomwolf — CHANGES_REQUESTED** [review]
Thanks a lot for all this work, it's great!

I've made a few comments on things to update. Mostly that we are only going to add `sacremoses` as required dependencies and raise error messages for the others.

I need to do a few modifications up-stream as mentioned in the comments to make it easier here.
Will do it in another PR so you can have a look.

File: pytorch_transformers/tokenization_xlm.py (hunk 1 above)
Comment (thomwolf) [reply pytorch_transformers/tokenization_xlm.py]: Sounds good to me indeed

**Comment by thomwolf** [comment]
Hi @shijie-wu,
So I've taken advantage of this PR to add a clean mechanism to set, save and reload tokenizer configurations.
This should fix in particular a recurring issue mentioned in #1158 and #1026 (failing to reload the lower casing configuration of the tokenizer) but more generally this is essential now for XLM more complex languages configuration.
Hope you don't mind me highjacking the PR.

File: pytorch_transformers/tokenization_xlm.py (hunk 1 above)
Comment (thomwolf) [reply pytorch_transformers/tokenization_xlm.py]: Ok added this

**Comment by thomwolf** [comment]
Ok I think this is good to go. Let's merge it.
